src/outputs/parthenon_opmd.cpp

```diff
@@ -0,0 +1,517 @@
+#include <sstream>
+#include <stdexcept>
+#include <string>
+#include <sys/types.h>
+#include <tuple>
+#include <vector>
+
+// Parthenon headers
+#include "Kokkos_Core_fwd.hpp"
+#include "basic_types.hpp"
+#include "coordinates/coordinates.hpp"
+#include "defs.hpp"
+#include "driver/driver.hpp"
+#include "globals.hpp"
+#include "interface/state_descriptor.hpp"
+#include "interface/variable_state.hpp"
+#include "mesh/mesh.hpp"
+#include "mesh/meshblock.hpp"
+#include "openPMD/Dataset.hpp"
+#include "openPMD/Datatype.hpp"
+#include "openPMD/IO/Access.hpp"
+#include "openPMD/Iteration.hpp"
+#include "openPMD/Mesh.hpp"
+#include "openPMD/Series.hpp"
+#include "openPMD/backend/MeshRecordComponent.hpp"
+#include "outputs/output_utils.hpp"
+#include "outputs/outputs.hpp"
+#include "outputs/parthenon_hdf5.hpp" // needd for VALId_VEC_TYPES -> move
+#include "parthenon_array_generic.hpp"
+#include "utils/error_checking.hpp"
+#include "utils/instrument.hpp"
+
+// OpenPMD headers
+#ifdef PARTHENON_ENABLE_OPENPMD
+#include <openPMD/openPMD.hpp>
+#endif // ifdef PARTHENON_ENABLE_OPENPMD
+
+namespace parthenon {
+
+using namespace OutputUtils;
+
+template <typename T>
+void WriteAllParamsOfType(std::shared_ptr<StateDescriptor> pkg, openPMD::Iteration *it) {
+  const std::string prefix = "Params/" + pkg->label() + "/";
+  const auto &params = pkg->AllParams();
+  for (const auto &key : params.GetKeys()) {
+    const auto type = params.GetType(key);
+    if (type == std::type_index(typeid(T))) {
+      // auto typed_ptr = dynamic_cast<Params::object_t<T> *>((p.second).get());
+      it->setAttribute(prefix + key, params.Get<T>(key));
+    }
+  }
+}
+
+template <typename... Ts>
+void WriteAllParamsOfMultipleTypes(std::shared_ptr<StateDescriptor> pkg,
+                                   openPMD::Iteration *it) {
+  ([&] { WriteAllParamsOfType<Ts>(pkg, it); }(), ...);
+}
+
+template <typename T>
+void WriteAllParams(std::shared_ptr<StateDescriptor> pkg, openPMD::Iteration *it) {
+  WriteAllParamsOfMultipleTypes<T, std::vector<T>>(pkg, it);
+  // TODO(pgrete) check why this doens't work, i.e., which type is causing problems
+  // WriteAllParamsOfMultipleTypes<PARTHENON_ATTR_VALID_VEC_TYPES(T)>(pkg, it);
+}
+
+namespace OpenPMDUtils {
+
+std::tuple<std::string, std::string> GetMeshRecordAndComponentNames(const VarInfo &vinfo,
+                                                                    const int comp_idx,
+                                                                    const int level) {
+  std::string comp_name;
+  if (vinfo.is_vector) {
+    if (comp_idx == 0) {
+      comp_name = "x";
+    } else if (comp_idx == 1) {
+      comp_name = "y";
+    } else if (comp_idx == 2) {
+      comp_name = "z";
+    } else {
+      PARTHENON_THROW("Expected component index doesn't match vector expectation.");
+    }
+    //  Current unclear how to properly handle other vectors and tensors, so everything
+    //  that not's a proper vector is a a scalar for now.
```

Comment:
Is it a scalar? or a collection of scalars? Does OpenPMD support array-valued fields?

Reply:
I'm still investigating.